File: app/src/main/java/mm/kso/movieapplication/viewmodels/HomeViewModel.java
package mm.kso.movieapplication.viewmodels;

import android.util.Log;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import java.util.ArrayList;
import java.util.HashMap;

import javax.inject.Inject;
import dagger.hilt.android.lifecycle.HiltViewModel;
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.functions.Function;
import io.reactivex.rxjava3.observers.DisposableObserver;
import io.reactivex.rxjava3.schedulers.Schedulers;
import mm.kso.movieapplication.model.Movie;
import mm.kso.movieapplication.model.MovieResponse;
import mm.kso.movieapplication.repository.Repository;

@HiltViewModel
public class HomeViewModel extends ViewModel {

    private static final String TAG = "HomeViewModel";
    private Repository repository;
    private MutableLiveData<ArrayList<Movie>> currentMoviesList = new MutableLiveData<>();
    private final CompositeDisposable disposable = new CompositeDisposable();
    private MutableLiveData<Movie> movieDetails = new MutableLiveData<>();
    private MutableLiveData<ArrayList<Movie>> popularMoviesList = new MutableLiveData<>();
    private MutableLiveData<ArrayList<Movie>> topRatedMoviesList = new MutableLiveData<>();
    private MutableLiveData<ArrayList<Movie>> upcomingMoviesList = new MutableLiveData<>();

    @Inject
    public HomeViewModel(Repository repository) {
        this.repository = repository;
    }

    public MutableLiveData<ArrayList<Movie>> getCurrentlyShowingList(){
        return currentMoviesList;
    }

    public MutableLiveData<ArrayList<Movie>> getTopRatedMoviesList(){
        return topRatedMoviesList;
    }

    public MutableLiveData<ArrayList<Movie>> getUpcomingMoviesList(){
        return upcomingMoviesList;
    }

    public MutableLiveData<Movie> getMovie() {
        return movieDetails;
    }

    public void getCurrentlyShowingMovies(HashMap<String, String> map){
        disposable.add(repository.getCurrentlyShowing(map)
                .subscribeOn(Schedulers.io())
                .map(new Function<MovieResponse, ArrayList<Movie>>() {
                    @Override
                    public ArrayList<Movie> apply(MovieResponse movieResponse) throws Throwable {
                        return movieResponse.getResults();
                    }
                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(new DisposableObserver<ArrayList<Movie>>() {
                    @Override
                    public void onNext(@io.reactivex.rxjava3.annotations.NonNull ArrayList<Movie> movies) {
                        currentMoviesList.setValue(movies);
                    }

                    @Override
                    public void onError(@io.reactivex.rxjava3.annotations.NonNull Throwable e) {

                    }

                    @Override
                    public void onComplete() {

                    }
                })
        );
    }

    public MutableLiveData<ArrayList<Movie>> getPopularMoviesList(){
        return popularMoviesList;
    }

    public void getPopularMovies(HashMap<String, String> map){
        disposable.add(repository.getPopular(map)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(result->popularMoviesList.setValue(result.getResults()),
                        error-> Log.e(TAG, "getPopularMovies: " + error.getMessage() ))
        );
    }

    public void getTopRatedMovies(HashMap<String, String> map) {
        disposable.add(repository.getTopRated(map)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(result -> topRatedMoviesList.setValue(result.getResults()),
                        error -> Log.e(TAG, "getTopRated: " + error.getMessage()))
        );
    }

    public void getUpcomingMovies(HashMap<String, String> map) {
        disposable.add(repository.getUpcoming(map)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(result -> upcomingMoviesList.setValue(result.getResults()),
                        error -> Log.e(TAG, "getUpcoming: " + error.getMessage()))
        );
    }


}

File: app/src/main/java/mm/kso/movieapplication/viewmodels/HomeViewModel.kt
package mm.kso.movieapplication.viewmodels

import android.util.Log
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import dagger.hilt.android.lifecycle.HiltViewModel
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.annotations.NonNull
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.observers.DisposableObserver
import io.reactivex.rxjava3.schedulers.Schedulers
import mm.kso.movieapplication.model.Movie
import mm.kso.movieapplication.model.MovieResponse
import mm.kso.movieapplication.repository.Repository
import java.util.*
import javax.inject.Inject

@HiltViewModel
class HomeViewModel @Inject constructor(private val repository: Repository) : ViewModel() {
    val currentlyShowingList = MutableLiveData<ArrayList<Movie>?>()
    private val disposable = CompositeDisposable()
    val movie = MutableLiveData<Movie>()
    val popularMoviesList = MutableLiveData<ArrayList<Movie>>()
    val topRatedMoviesList = MutableLiveData<ArrayList<Movie>>()
    val upcomingMoviesList = MutableLiveData<ArrayList<Movie>>()
    fun getCurrentlyShowingMovies(map: HashMap<String, String>) {
        disposable.add(
            repository.getCurrentlyShowing(map)
                ?.subscribeOn(Schedulers.io())
                ?.map { movieResponse -> movieResponse?.results }
                ?.observeOn(AndroidSchedulers.mainThread())
                ?.subscribeWith(object : DisposableObserver<ArrayList<Movie>>() {
                    override fun onNext(movies: @NonNull ArrayList<Movie>?) {
                        currentlyShowingList.setValue(movies)
                    }

                    override fun onError(e: @NonNull Throwable?) {}
                    override fun onComplete() {}
                })
        )
    }

    fun getPopularMovies(map: HashMap<String, String>) {
        disposable.add(
            repository.getPopular(map)
                ?.subscribeOn(Schedulers.io())
                ?.observeOn(AndroidSchedulers.mainThread())
                ?.subscribe(
                    { result: MovieResponse? -> popularMoviesList.setValue(result?.results) }
                ) { error: Throwable -> Log.e(TAG, "getPopularMovies: " + error.message) }
        )
    }

    fun getTopRatedMovies(map: HashMap<String, String>) {
        disposable.add(
            repository.getTopRated(map)
                ?.subscribeOn(Schedulers.io())
                ?.observeOn(AndroidSchedulers.mainThread())
                ?.subscribe(
                    { result: MovieResponse? -> topRatedMoviesList.setValue(result?.results) }
                ) { error: Throwable -> Log.e(TAG, "getTopRated: " + error.message) }
        )
    }

    fun getUpcomingMovies(map: HashMap<String, String>) {
        disposable.add(
            repository.getUpcoming(map)
                ?.subscribeOn(Schedulers.io())
                ?.observeOn(AndroidSchedulers.mainThread())
                ?.subscribe(
                    { result: MovieResponse? -> upcomingMoviesList.setValue(result?.results) }
                ) { error: Throwable -> Log.e(TAG, "getUpcoming: " + error.message) }
        )
    }

    companion object {
        private const val TAG = "HomeViewModel"
    }
}